rasa/core/actions/action.py

```diff
@@ -277,16 +293,61 @@ async def run(
         return [create_bot_utterance(message)]
 
     def name(self) -> Text:
+        """Returns action name."""
         return self.template_name
 
-    def __str__(self) -> Text:
-        return "ActionUtterTemplate('{}')".format(self.name())
+
+class ActionEndToEndResponse(Action):
+    """Action to utter end-to-end responses to the user."""
+
+    def __init__(self, action_text: Text) -> None:
+        """Creates action.
+
+        Args:
+            action_text: Text of end-to-end bot response.
+        """
+        self.action_text = action_text
+
+    def name(self) -> Text:
+        """Returns action name."""
+        # In case of an end-to-end action there is no label (aka name) for the action.
```

Comment:
do we predict actual text in the model or do we predict a label (even if not explicitly) by selecting one of the known and used responses in the training data? if so, why not assign them names?

Reply:
we predict the actual text as the label. but text is used for featurization, similar as in old response selector. If we assign labels, we add unneeded complexity, and if we use these labels for features, we break the idea of e2e

Reply:
I don't think it is a good idea to rely on the name as an identifier:
- what happens if there are buttons?
- what if the response is just an image?
- what if you have a stored conversation and want to change the bots text?

I don't yet understand what the benefit of the work-around of storing text on the action is compared to introducing a generated name. Responses could be stored on the domain and you'd still be able to retrieve the response to do the prediction using the actual text instead of the label.
If

Reply:
this is impossible in case of e2e. In the cases above you should use templates

Reply:
the idea is to have a capability to have full text-text model. I think it is good if internal representation reflect that. It makes it easier to debug later

Reply:
I think we still need to keep in mind that we need to extend this to support the whole message format, not just text.